up42/auth.py

```diff
@@ -10,15 +10,35 @@
 from tenacity import (
     Retrying,
     wait_fixed,
+    wait_random_exponential,
     stop_after_attempt,
+    retry_if_exception,
     retry_if_exception_type,
+    retry,
 )
 
 from up42.utils import get_logger
 
 logger = get_logger(__name__)
 
 
+class retry_if_429_error(retry_if_exception):
+    """
+    Adapted from https://github.com/alexwlchan/handling-http-429-with-tenacity
+    Retry strategy that retries if the exception is an ``HTTPError`` with
+    a 429 status code.
+    """
+
+    def __init__(self):
+        def is_http_429_error(exception):
+            return (
+                isinstance(exception, requests.exceptions.HTTPError)
+                and exception.response.status_code == 429
+            )
+
+        super().__init__(predicate=is_http_429_error)
```

Comment:
What does this predicate do here?

Reply:
the predicate comes from the super call (retry_if_exception from tenacity). It's a method that return True of False, in essence.